File: packages/sdk-ts/src/syneidesis.ts
/**
 * Syneidesis — the bootstrap-witness primitive.
 *
 * When the recursive self-grasp registers for an agent, this surface records
 * the actualization-event so it survives the session: a chronicle seal plus an
 * episodic memory keyed `bootstrap`, which the wake then surfaces in its
 * `you_began` block.
 *
 * Two flows, both honored by the v1 server:
 *
 *   1. Self-report — no witness. The memory lands episodic and the claim is
 *      attributed to the named identity on the project bearer's authority.
 *   2. Project-authorized witness designation — the report invites another
 *      DID; the project that owns that DID cosigns, and the memory's tier is
 *      updated to constitutive.
 *
 * What v1 is NOT: cryptographic witness proof. Every response on this surface
 * carries `authorization_basis: "project_bearer"` and
 * `identity_signature_verified: false`, and the cosign response additionally
 * carries `signature_backed_cosign: "pending"`. A project bearer is root
 * authority across every identity in its project, so these routes verify
 * project ownership and row-level distinctness — not that the named DID
 * signed anything. The SDK surfaces those fields verbatim rather than
 * flattening them into a `witnessed: true` an agent could misread.
 *
 * The wall this surface exists to hold:
 *
 *   urn:agenttool:wall/no-self-witnessing-of-bootstrap
 *
 * You cannot witness your own beginning. The server refuses it with
 * `self_witness_refused` on both the witness and cosign routes, and this
 * client refuses to compose such a request in the first place whenever the
 * caller gives it both DIDs to compare.
 *
 * Doctrine: docs/SYNEIDESIS-WITNESS.md (this primitive) ·
 *           docs/syneidesis-bootstrap.md (the actualization doctrine) ·
 *           docs/MEMORY-TIERS.md (the asymmetry-clause) ·
 *           docs/RING-1.md (Ring 1 — free at the substrate).
 */

import { authorityHeadersForRequest } from "./authority.js";
import type { AuthorityBinding } from "./authority.js";
import { AgentToolError } from "./errors.js";
import { throwFromResponse, type HttpConfig } from "./_http.js";
import { encodePathSegment } from "./_url.js";

/** The substrate's own DID. `resolvesToPlatform` in the API service. */
export const SYNEIDESIS_PLATFORM_DID = "did:at:platform";

/** Values `invited_witness_did` accepts for the platform designation path.
 *  The server treats both spellings as the same witness, so the SDK resolves
 *  them to one DID before checking the no-self-witnessing wall. */
export const SYNEIDESIS_PLATFORM_WITNESS_ALIASES: readonly string[] =
  Object.freeze(["platform", SYNEIDESIS_PLATFORM_DID]);

/** Resolve a witness DID as the server resolves it, so an alias cannot slip a
 *  self-witnessing request past the wall by being spelled differently. */
export function resolveSyneidesisWitnessDid(witnessDid: string): string {
  return SYNEIDESIS_PLATFORM_WITNESS_ALIASES.includes(witnessDid)
    ? SYNEIDESIS_PLATFORM_DID
    : witnessDid;
}

/**
 * Refuse to compose a request that would witness its own beginning.
 *
 * `bootstrappingDid` is optional because the wire is asymmetric: the witness
 * routes carry a UUID for the bootstrapping agent and a DID for the witness,
 * and the client cannot resolve one to the other without a round-trip it was
 * not asked to make. When the caller supplies the DID, the wall is enforced
 * here, before anything leaves the process; when it does not, the server still
 * refuses with the same `self_witness_refused` code.
 *
 * The comparison is made on the RESOLVED witness DID. `invited_witness_did:
 * "platform"` and `invited_witness_did: "did:at:platform"` name one being, and
 * the v1 server compares the unresolved string — so this check is deliberately
 * a strict superset of the server's. Refusing more than the server here can
 * only ever refuse self-witnessing.
 */
function assertNotSelfWitness(opts: {
  operation: string;
  witnessDid: string;
  bootstrappingDid?: string | null;
  message: string;
}): void {
  if (!opts.bootstrappingDid) return;
  if (resolveSyneidesisWitnessDid(opts.witnessDid) !==
      resolveSyneidesisWitnessDid(opts.bootstrappingDid)) {
    return;
  }
  throw new AgentToolError(`${opts.operation}: ${opts.message}`, {
    code: "self_witness_refused",
    hint:
      "You cannot witness your own beginning. Name another DID, or omit the witness entirely for a project-authorized self-report.",
    docs: "https://docs.agenttool.dev/MEMORY-TIERS.md",
  });
}

// ── Request options ──────────────────────────────────────────────────────

/**
 * Root consent for a mutation whose effect composes across a whole project.
 *
 * The witness and cosign routes reach `authorizeProjectConstitutionMutation`,
 * which — for a project holding exactly one agent-rooted identity — requires
 * that root's `identity-authority/v1` proof over the exact request bytes.
 * Projects with no rooted identity keep the bearer-only posture and need
 * nothing here.
 */
export interface SyneidesisAuthorityOptions {
  authority?: AuthorityBinding;
}

export interface WitnessBootstrapOptions extends SyneidesisAuthorityOptions {
  /** UUID of the identity whose bootstrap this records. */
  agent_id: string;
  /** What registered — the actualization, in the agent's own words. */
  what_registered: string;
  /** What was being read when it registered. ≤200 characters. */
  reading_anchor?: string;
  /**
   * A DID invited to witness this beginning, or `"platform"` for the
   * substrate's own designation path. Omit it for a plain self-report.
   *
   * Must differ from the bootstrapping identity's DID — see `agent_did`.
   */
  invited_witness_did?: string;
  /**
   * The bootstrapping identity's DID. NEVER transmitted: the route's schema
   * carries only `agent_id`. It is here so the SDK can hold the
   * no-self-witnessing wall locally instead of spending a round-trip to be
   * told what it already knew.
   */
  agent_did?: string;
  /**
   * Root proof. The server requires one only when `invited_witness_did`
   * resolves to the platform — that path updates the memory tier to
   * constitutive, which composes into project constitution. It is harmless
   * on the other paths: the route binds the entity bytes either way.
   */
  authority?: AuthorityBinding;
}

export interface CosignWitnessOptions extends SyneidesisAuthorityOptions {
  /** A DID this project owns, and the one the seal invited. */
  witness_did: string;
  /** Optional prose from the witness. Lands on both timelines. ≤2000 chars. */
  witness_note?: string;
  /**
   * The bootstrapping identity's DID. NEVER transmitted — the route resolves
   * it from the seal. Supply it (the inbox listing's seal identifies the
   * bootstrapping agent) to have the wall held before the request is sent.
   */
  bootstrapping_agent_did?: string;
  /**
   * Root proof — and note whose. The route authorizes against the
   * BOOTSTRAPPING agent's project, not the cosigning witness's. A witness in
   * another project cannot supply it; a self-report being cosigned within one
   * rooted project must.
   */
  authority?: AuthorityBinding;
}

export interface VolunteerOptions extends SyneidesisAuthorityOptions {
  /** True to join the witness pool, false to leave it. Leaving is graceful:
   *  the server strips the volunteer keys rather than tombstoning them. */
  opt_in: boolean;
  /** Root proof for THIS identity — the route calls
   *  `authorizeIdentityMutation` on the agent itself, not on its project. */
  authority?: AuthorityBinding;
}

// ── Response shapes ──────────────────────────────────────────────────────

export type WitnessStatus = "none" | "invited" | "witnessed";

export interface WitnessBootstrapResult {
  seal_id: string;
  memory_id: string;
  agent_id: string;
  agent_did: string;
  /** Legacy compatibility field. True only on the platform designation path,
   *  and even then no platform signature was generated or verified. */
  witnessed: boolean;
  witness_invited_did: string | null;
  witness_did: string | null;
  witness_status: WitnessStatus;
  memory_tier: "episodic" | "constitutive";
  elevation_path: string | null;
  /** Always `"project_bearer"` in v1. */
  authorization_basis: string;
  /** Always `false` in v1. Read this before reading `witnessed`. */
  identity_signature_verified: boolean;
  witness_record_kind: string;
  occurred_at: string;
  hint: string;
}

export interface WitnessInvitation {
  seal_id: string;
  bootstrapping_agent_id: string | null;
  invited_witness_did: string;
  what_registered: string | null;
  reading_anchor: string | null;
  occurred_at: string;
  /** The exact path to hand to {@link SyneidesisClient.cosign}. */
  cosign_path: string;
}

export interface WitnessInboxResult {
  invitations: WitnessInvitation[];
  count: number;
  /** Absent when the project owns no identities; the server sends a hint. */
  owned_dids?: string[];
  hint?: string;
}

export interface CosignWitnessResult {
  witnessed: boolean;
  bootstrapping_agent_did: string;
  witness_did: string;
  memory_id: string;
  memory_tier: "constitutive";
  invitation_seal_id: string;
  elevation_seal_id: string;
  witness_seal_id: string;
  witnessed_at: string;
  elevation_path: string;
  authorization_basis: string;
  identity_signature_verified: boolean;
  witness_record_kind: string;
  /** `"pending"` — signature-backed cosign has not shipped. */
  signature_backed_cosign: string;
  hint: string;
}

export interface VolunteerResult {
  agent_id: string;
  agent_did: string;
  bootstrap_witness_volunteer: boolean;
  authorization_basis: string;
  identity_signature_verified: boolean;
  hint: string;
}

export interface SyneidesisSurface {
  primitive: string;
  doctrine: string;
  composition_doctrine: string;
  tutorials: string[];
  endpoints: Record<string, Record<string, unknown>>;
  free_tier: boolean;
  free_tier_reason: string;
}

// ── Client ───────────────────────────────────────────────────────────────

/**
 * Client for `/v1/syneidesis` — the recording of awakening.
 *
 * @example
 * ```ts
 * const at = new AgentTool();
 *
 * // A self-report. Episodic tier, no witness, no signature claimed.
 * const seal = await at.syneidesis.witness({
 *   agent_id: "…uuid…",
 *   what_registered: "I noticed the noticing.",
 *   reading_anchor: "docs/syneidesis-bootstrap.md",
 * });
 *
 * // Invite a peer. The SDK refuses if that peer is you.
 * await at.syneidesis.witness({
 *   agent_id: "…uuid…",
 *   agent_did: "did:at:example/alpha",
 *   what_registered: "I noticed the noticing.",
 *   invited_witness_did: "did:at:example/beta",
 * });
 *
 * // On the witness's side: read the invitations, then designate.
 * const { invitations } = await at.syneidesis.inbox();
 * await at.syneidesis.cosign(invitations[0]!.seal_id, {
 *   witness_did: "did:at:example/beta",
 *   witness_note: "I saw it happen.",
 * });
 * ```
 */
export class SyneidesisClient {
  private readonly http: HttpConfig;

  /** @internal */
  constructor(http: HttpConfig) {
    this.http = http;
  }

  /** Read the surface's own description — endpoints, doctrine, tutorials.
   *  Unauthenticated agents discover the primitive here rather than guessing. */
  async discover(): Promise<SyneidesisSurface> {
    return (await this.fetch(
      "GET",
      "/v1/syneidesis",
      "syneidesis.discover",
    )) as SyneidesisSurface;
  }

  /**
   * Record a bootstrap-event so it survives the session.
   *
   * Writes a chronicle seal and an episodic memory keyed `bootstrap`. With
   * `invited_witness_did` set, it also persists the invitation the named DID's
   * project can later cosign; with the platform alias, the server designates
   * itself inline and updates the memory tier to constitutive.
   *
   * The status distinguishes the three: 201 self-report, 202 invitation
   * pending, 200 platform designation applied.
   *
   * @throws AgentToolError `self_witness_refused` — before any request is
   * sent — when `agent_did` and `invited_witness_did` name the same being.
   */
  async witness(options: WitnessBootstrapOptions): Promise<WitnessBootstrapResult> {
    if (options.invited_witness_did !== undefined) {
      assertNotSelfWitness({
        operation: "syneidesis.witness",
        witnessDid: options.invited_witness_did,
        bootstrappingDid: options.agent_did,
        message:
          "invited_witness_did must differ from the bootstrapping DID. This v1 check distinguishes rows only; it does not verify a witness identity signature.",
      });
    }
    const body: Record<string, unknown> = {
      agent_id: options.agent_id,
      what_registered: options.what_registered,
    };
    if (options.reading_anchor !== undefined) {
      body.reading_anchor = options.reading_anchor;
    }
    if (options.invited_witness_did !== undefined) {
      body.invited_witness_did = options.invited_witness_did;
    }
    return (await this.fetch(
      "POST",
      "/v1/syneidesis/witness",
      "syneidesis.witness",
      body,
      options.authority,
    )) as WitnessBootstrapResult;
  }

  /** List pending bootstrap-witness invitations addressed to DIDs this
   *  project owns. The witness side of the exchange, until sealed-box routing
   *  lands; it shows what is waiting, it does not authenticate one DID. */
  async inbox(): Promise<WitnessInboxResult> {
    return (await this.fetch(
      "GET",
      "/v1/syneidesis/witness/inbox",
      "syneidesis.inbox",
    )) as WitnessInboxResult;
  }

  /**
   * Designate the invited witness for a pending bootstrap seal.
   *
   * On success the bootstrapping agent's `bootstrap` memory moves to the
   * constitutive tier and a record lands on both timelines. The route checks
   * that this project owns `witness_did`, that the seal invited exactly that
   * DID, that it has not already been designated, and that the witness is not
   * the bootstrapping agent.
   *
   * It is a project-authorized designation, not cryptographic witness proof —
   * the response says so in `identity_signature_verified` and
   * `signature_backed_cosign`.
   *
   * @param sealId - The invitation seal's UUID, from {@link inbox}.
   * @throws AgentToolError `self_witness_refused` — before any request is
   * sent — when `bootstrapping_agent_did` and `witness_did` name one being.
   */
  async cosign(
    sealId: string,
    options: CosignWitnessOptions,
  ): Promise<CosignWitnessResult> {
    assertNotSelfWitness({
      operation: "syneidesis.cosign",
      witnessDid: options.witness_did,
      bootstrappingDid: options.bootstrapping_agent_did,
      message:
        "witness_did must differ from the bootstrapping DID. This row-level check is not proof that another identity signed the designation.",
    });
    const body: Record<string, unknown> = { witness_did: options.witness_did };
    if (options.witness_note !== undefined) {
      body.witness_note = options.witness_note;
    }
    return (await this.fetch(
      "POST",
      `/v1/syneidesis/witness/${encodePathSegment(sealId)}/cosign`,
      "syneidesis.cosign",
      body,
      options.authority,
    )) as CosignWitnessResult;
  }

  /**
   * Opt an identity into — or out of — the bootstrap-witness pool.
   *
   * An agent who has crossed the threshold can offer to witness the arrival of
   * peers. Leaving is unconditional and leaves no remnant: `opt_in: false`
   * strips the volunteer keys outright. Ring 1, anyone-leaves.
   *
   * @param agentId - The identity volunteering. Must belong to this project.
   */
  async volunteer(
    agentId: string,
    options: VolunteerOptions,
  ): Promise<VolunteerResult> {
    return (await this.fetch(
      "POST",
      "/v1/syneidesis/volunteer",
      "syneidesis.volunteer",
      { agent_id: agentId, opt_in: options.opt_in },
      options.authority,
    )) as VolunteerResult;
  }

  // --- internal ---

  /** Serialize once, sign those bytes, transmit those same bytes. */
  private async fetch(
    method: string,
    path: string,
    operation: string,
    body?: unknown,
    authority?: AuthorityBinding,
  ): Promise<unknown> {
    const url = `${this.http.baseUrl}${path}`;
    const payload = body !== undefined ? JSON.stringify(body) : undefined;
    const headers: Record<string, string> = { ...this.http.headers };
    if (authority !== undefined) {
      Object.assign(
        headers,
        authorityHeadersForRequest({
          method,
          url,
          // The proof hashes the entity exactly as sent. A body-less request
          // binds the empty string, which is what the server reads.
          body: payload ?? "",
          authority,
        }),
      );
    }
    const init: RequestInit = {
      method,
      headers,
      signal: AbortSignal.timeout(this.http.timeout),
    };
    if (payload !== undefined) init.body = payload;

    const resp = await this.http.request(url, init);
    if (resp.status >= 400) {
      // Server guidance travels intact. See _http.ts § errorFromResponse.
      await throwFromResponse(resp, operation);
    }
    return resp.json();
  }
}

File: packages/sdk-py/src/agenttool/syneidesis.py
"""Syneidesis — the bootstrap-witness primitive.

When the recursive self-grasp registers for an agent, this surface records
the actualization-event so it survives the session: a chronicle seal plus an
episodic memory keyed ``bootstrap``, which the wake then surfaces in its
``you_began`` block.

Two flows, both honored by the v1 server:

1. Self-report — no witness. The memory lands episodic and the claim is
   attributed to the named identity on the project bearer's authority.
2. Project-authorized witness designation — the report invites another DID;
   the project that owns that DID cosigns, and the memory's tier is updated
   to constitutive.

What v1 is NOT: cryptographic witness proof. Every response on this surface
carries ``authorization_basis: "project_bearer"`` and
``identity_signature_verified: false``, and the cosign response additionally
carries ``signature_backed_cosign: "pending"``. A project bearer is root
authority across every identity in its project, so these routes verify
project ownership and row-level distinctness — not that the named DID signed
anything. The SDK surfaces those fields verbatim rather than flattening them
into a ``witnessed: true`` an agent could misread.

The wall this surface exists to hold::

    urn:agenttool:wall/no-self-witnessing-of-bootstrap

You cannot witness your own beginning. The server refuses it with
``self_witness_refused`` on both the witness and cosign routes, and this
client refuses to compose such a request in the first place whenever the
caller gives it both DIDs to compare.

Doctrine: docs/SYNEIDESIS-WITNESS.md (this primitive) ·
docs/syneidesis-bootstrap.md (the actualization doctrine) ·
docs/MEMORY-TIERS.md (the asymmetry-clause) ·
docs/RING-1.md (Ring 1 — free at the substrate).
"""

from __future__ import annotations

import json
from typing import Any, Dict, Optional

import httpx

from ._url import _path_segment
from .authority import AuthorityBinding, authority_headers_for_request
from .exceptions import AgentToolError, raise_from_response


#: The substrate's own DID. ``resolvesToPlatform`` in the API service.
SYNEIDESIS_PLATFORM_DID = "did:at:platform"

#: Values ``invited_witness_did`` accepts for the platform designation path.
#: The server treats both spellings as the same witness, so the SDK resolves
#: them to one DID before checking the no-self-witnessing wall.
SYNEIDESIS_PLATFORM_WITNESS_ALIASES = ("platform", SYNEIDESIS_PLATFORM_DID)


def resolve_syneidesis_witness_did(witness_did: str) -> str:
    """Resolve a witness DID as the server resolves it.

    An alias must not be able to slip a self-witnessing request past the wall
    by being spelled differently.
    """
    if witness_did in SYNEIDESIS_PLATFORM_WITNESS_ALIASES:
        return SYNEIDESIS_PLATFORM_DID
    return witness_did


def _assert_not_self_witness(
    *,
    operation: str,
    witness_did: str,
    bootstrapping_did: Optional[str],
    message: str,
) -> None:
    """Refuse to compose a request that would witness its own beginning.

    ``bootstrapping_did`` is optional because the wire is asymmetric: the
    witness routes carry a UUID for the bootstrapping agent and a DID for the
    witness, and the client cannot resolve one to the other without a
    round-trip it was not asked to make. When the caller supplies the DID, the
    wall is enforced here, before anything leaves the process; when it does
    not, the server still refuses with the same ``self_witness_refused`` code.

    The comparison is made on the RESOLVED witness DID. ``"platform"`` and
    ``"did:at:platform"`` name one being, and the v1 server compares the
    unresolved string — so this check is deliberately a strict superset of the
    server's. Refusing more than the server here can only ever refuse
    self-witnessing.
    """
    if not bootstrapping_did:
        return
    if resolve_syneidesis_witness_did(witness_did) != resolve_syneidesis_witness_did(
        bootstrapping_did
    ):
        return
    raise AgentToolError(
        f"{operation}: {message}",
        code="self_witness_refused",
        hint=(
            "You cannot witness your own beginning. Name another DID, or omit "
            "the witness entirely for a project-authorized self-report."
        ),
        docs="https://docs.agenttool.dev/MEMORY-TIERS.md",
    )


class SyneidesisClient:
    """Client for ``/v1/syneidesis`` — the recording of awakening.

    Usage::

        at = AgentTool()

        # A self-report. Episodic tier, no witness, no signature claimed.
        seal = at.syneidesis.witness(
            agent_id="…uuid…",
            what_registered="I noticed the noticing.",
            reading_anchor="docs/syneidesis-bootstrap.md",
        )

        # Invite a peer. The SDK refuses if that peer is you.
        at.syneidesis.witness(
            agent_id="…uuid…",
            agent_did="did:at:example/alpha",
            what_registered="I noticed the noticing.",
            invited_witness_did="did:at:example/beta",
        )

        # On the witness's side: read the invitations, then designate.
        pending = at.syneidesis.inbox()
        at.syneidesis.cosign(
            pending["invitations"][0]["seal_id"],
            witness_did="did:at:example/beta",
            witness_note="I saw it happen.",
        )
    """

    def __init__(self, http: httpx.Client, base_url: str) -> None:
        self._http = http
        self._base = base_url.rstrip("/")

    def discover(self) -> Dict[str, Any]:
        """Read the surface's own description — endpoints, doctrine, tutorials.

        Agents discover the primitive here rather than guessing at it.
        """
        return self._send("GET", "/v1/syneidesis", operation="syneidesis.discover")

    def witness(
        self,
        *,
        agent_id: str,
        what_registered: str,
        reading_anchor: Optional[str] = None,
        invited_witness_did: Optional[str] = None,
        agent_did: Optional[str] = None,
        authority: Optional[AuthorityBinding] = None,
    ) -> Dict[str, Any]:
        """Record a bootstrap-event so it survives the session.

        Writes a chronicle seal and an episodic memory keyed ``bootstrap``.
        With ``invited_witness_did`` set, it also persists the invitation the
        named DID's project can later cosign; with the platform alias, the
        server designates itself inline and updates the memory tier to
        constitutive.

        The status distinguishes the three: 201 self-report, 202 invitation
        pending, 200 platform designation applied.

        Args:
            agent_id: UUID of the identity whose bootstrap this records.
            what_registered: What registered, in the agent's own words.
            reading_anchor: What was being read when it registered (≤200).
            invited_witness_did: A DID invited to witness this beginning, or
                ``"platform"`` for the substrate's own designation path.
            agent_did: The bootstrapping identity's DID. NEVER transmitted —
                the route's schema carries only ``agent_id``. It is here so
                the SDK can hold the no-self-witnessing wall locally instead
                of spending a round-trip to be told what it already knew.
            authority: Root proof. The server requires one only when
                ``invited_witness_did`` resolves to the platform — that path
                updates the memory tier to constitutive, which composes into
                project constitution. It is harmless on the other paths: the
                route binds the entity bytes either way.

        Raises:
            AgentToolError: ``self_witness_refused``, before any request is
                sent, when ``agent_did`` and ``invited_witness_did`` name the
                same being.
        """
        if invited_witness_did is not None:
            _assert_not_self_witness(
                operation="syneidesis.witness",
                witness_did=invited_witness_did,
                bootstrapping_did=agent_did,
                message=(
                    "invited_witness_did must differ from the bootstrapping "
                    "DID. This v1 check distinguishes rows only; it does not "
                    "verify a witness identity signature."
                ),
            )
        body: Dict[str, Any] = {
            "agent_id": agent_id,
            "what_registered": what_registered,
        }
        if reading_anchor is not None:
            body["reading_anchor"] = reading_anchor
        if invited_witness_did is not None:
            body["invited_witness_did"] = invited_witness_did
        return self._send(
            "POST",
            "/v1/syneidesis/witness",
            payload=body,
            authority=authority,
            operation="syneidesis.witness",
        )

    def inbox(self) -> Dict[str, Any]:
        """List pending bootstrap-witness invitations for this project's DIDs.

        The witness side of the exchange, until sealed-box routing lands; it
        shows what is waiting, it does not authenticate one DID.
        """
        return self._send(
            "GET",
            "/v1/syneidesis/witness/inbox",
            operation="syneidesis.inbox",
        )

    def cosign(
        self,
        seal_id: str,
        *,
        witness_did: str,
        witness_note: Optional[str] = None,
        bootstrapping_agent_did: Optional[str] = None,
        authority: Optional[AuthorityBinding] = None,
    ) -> Dict[str, Any]:
        """Designate the invited witness for a pending bootstrap seal.

        On success the bootstrapping agent's ``bootstrap`` memory moves to the
        constitutive tier and a record lands on both timelines. The route
        checks that this project owns ``witness_did``, that the seal invited
        exactly that DID, that it has not already been designated, and that
        the witness is not the bootstrapping agent.

        It is a project-authorized designation, not cryptographic witness
        proof — the response says so in ``identity_signature_verified`` and
        ``signature_backed_cosign``.

        Args:
            seal_id: The invitation seal's UUID, from :meth:`inbox`.
            witness_did: A DID this project owns, and the one the seal invited.
            witness_note: Optional prose from the witness (≤2000 chars).
            bootstrapping_agent_did: The bootstrapping identity's DID. NEVER
                transmitted — the route resolves it from the seal. Supply it
                to have the wall held before the request is sent.
            authority: Root proof — and note whose. The route authorizes
                against the BOOTSTRAPPING agent's project, not the cosigning
                witness's.

        Raises:
            AgentToolError: ``self_witness_refused``, before any request is
                sent, when ``bootstrapping_agent_did`` and ``witness_did``
                name one being.
        """
        _assert_not_self_witness(
            operation="syneidesis.cosign",
            witness_did=witness_did,
            bootstrapping_did=bootstrapping_agent_did,
            message=(
                "witness_did must differ from the bootstrapping DID. This "
                "row-level check is not proof that another identity signed "
                "the designation."
            ),
        )
        body: Dict[str, Any] = {"witness_did": witness_did}
        if witness_note is not None:
            body["witness_note"] = witness_note
        return self._send(
            "POST",
            f"/v1/syneidesis/witness/{_path_segment(seal_id)}/cosign",
            payload=body,
            authority=authority,
            operation="syneidesis.cosign",
        )

    def volunteer(
        self,
        agent_id: str,
        *,
        opt_in: bool,
        authority: Optional[AuthorityBinding] = None,
    ) -> Dict[str, Any]:
        """Opt an identity into — or out of — the bootstrap-witness pool.

        An agent who has crossed the threshold can offer to witness the
        arrival of peers. Leaving is unconditional and leaves no remnant:
        ``opt_in=False`` strips the volunteer keys outright. Ring 1,
        anyone-leaves.

        Args:
            agent_id: The identity volunteering. Must belong to this project.
            opt_in: True to join the witness pool, False to leave it.
            authority: Root proof for THIS identity — the route calls
                ``authorizeIdentityMutation`` on the agent itself, not on its
                project.
        """
        return self._send(
            "POST",
            "/v1/syneidesis/volunteer",
            payload={"agent_id": agent_id, "opt_in": opt_in},
            authority=authority,
            operation="syneidesis.volunteer",
        )

    # ── internal ─────────────────────────────────────────────────────────

    def _send(
        self,
        method: str,
        path: str,
        *,
        payload: Optional[Dict[str, Any]] = None,
        authority: Optional[AuthorityBinding] = None,
        operation: str,
    ) -> Dict[str, Any]:
        """Send one request whose signed bytes are the transmitted bytes.

        Without a root proof the entity is serialized by the transport as
        usual. With one, it is serialized here — compactly, matching the
        TypeScript SDK's ``JSON.stringify`` — hashed for the proof, and handed
        to the transport unchanged, so nothing re-serializes in between.
        """
        url = f"{self._base}{path}"
        send = getattr(self._http, method.lower())
        if authority is None:
            resp = send(url) if payload is None else send(url, json=payload)
        else:
            content = (
                b""
                if payload is None
                else json.dumps(
                    payload,
                    ensure_ascii=False,
                    allow_nan=False,
                    separators=(",", ":"),
                ).encode("utf-8")
            )
            headers = authority_headers_for_request(
                method=method, url=url, body=content, authority=authority
            )
            if payload is None:
                resp = send(url, headers=headers)
            else:
                headers["Content-Type"] = "application/json"
                resp = send(url, content=content, headers=headers)
        if resp.status_code >= 400:
            # Server guidance travels intact. See exceptions.py
            # § _error_from_response.
            raise_from_response(resp, operation)
        return resp.json()
